### ISSUE TITLE
[feature request] Enforce Dangling Commas (comma-dangle) only in multiline array/object literals

### ISSUE THREAD

**Issue by UltCombo** [eslint/eslint#1984]
As title says, I'd like this code to lint:

```js
var arr = [
    1,
    2,
];
var obj = {
    a: 1,
    b: 2,
};
var arr2 = [1, 2];
var obj2 = { a: 1, b: 2 };
```

And the following patterns to fail linting:

```js
var arr = [
    1,
    2
];
var obj = {
    a: 1,
    b: 2
};
var arr2 = [1, 2,];
var obj2 = { a: 1, b: 2, };
```

Is this possible to implement as an option to `comma-dangle`?

Thanks.

**Comment by nzakas** [comment]
This isn't possible today, but shouldn't be too hard to add.

**Comment by UltCombo** [comment]
Thanks for the quick triaging.

I've been thinking about more complex use cases (e.g. mixed style, multiline array/object literals with some properties in the same line), and I believe the logic could be kept rather simple: only allow (and require) a trailing comma when there is a newline between the last item/property and the closing `]`/`}` respectively.

WDYT?

**Comment by nzakas** [comment]
I'm a bit confused by that wording. What I was thinking was to have an option "always-multiline" that enforces dangling commas only when the array or object literal is on multiple lines.

**Comment by UltCombo** [comment]
> What I was thinking was to have an option "always-multiline" that enforces dangling commas only when the array or object literal is on multiple lines.

Yes, that wording is clearer indeed. `=]`

**Comment by keithamus** [comment]
Just FYI, [JSCS offers this feature](http://jscs.info/rules.html#requiretrailingcomma) as `"requireTrailingComma": { "ignoreSingleLine": true }`

Similarly there is a plugin available in the eslint ecosystem; [eslint-plugin-require-trailing-comma](https://github.com/greggman/eslint-plugin-require-trailing-comma). This plugin has the following modes:

```yml
# No trailing comma is required
require-trailing-comma/require-trailing-comma: 0

# Multiline Objects or Arrays must have a trailing comma
require-trailing-comma/require-trailing-comma: 2

# Every Object or Array (including single lines) must have a trailing comma
require-trailing-comma/require-trailing-comma: [2, "all"]
```

As a workaround for now - you can simply set `comma-dangle: 0` and use [eslint-plugin-require-trailing-comma](https://github.com/greggman/eslint-plugin-require-trailing-comma) instead.

**Comment by UltCombo** [comment]
Woah, very nice! Thanks @keithamus.
Should this issue be closed, as there is already a plugin for this functionality? Or do you guys still want to land it on the core?

**Comment by nzakas** [comment]
We will still make a change. Just need to figure out the right change to make.

**Comment by keithamus** [comment]
My opinion is the following makes the most sense:

 - `comma-dangle: 0` : No enforcement.
    - Valid:

    ```js
    [1,2]
    [1,2,]
    {
        foo: 'bar'
    }
    {
        foo: 'bar',
    }
    ```
 - `comma-dangle: [2, "always"]` : Failure when dangling comma isnt present
    - Valid:
    ```js
    [1,2,]
    {
        foo: 'bar',
    }
    ```
   - Invalid:
    ```js
    [1,2]
    {
        foo: 'bar'
    }
    ```

 - `comma-dangle: [2, "multiline"]` : Failure when dangling is present on single lines, failure when not present on multilines:
    - Valid:
    ```js
    [1,2]
    {
        foo: 'bar',
    }
    ```
   - Invalid:
    ```js
    [1,2,]
    {
        foo: 'bar'
    }
    ```

 - `comma-dangle: [2, "never"]` : Failure whenever a dangling comma is present:
    - Valid:
    ```js
    [1,2]
    {
        foo: 'bar'
    }
    ```
   - Invalid:
    ```js
    [1,2,]
    {
        foo: 'bar',
    }
    ```

As far as I can see - this is pretty much exactly how the existing behaviour is implemented, with the exception of "multiline". 

I'm keen to get this in - and so if everyone is happy with the above proposal (effectively adding the "multiline" option), I'll formulate a PR for this ASAP.

**Comment by UltCombo** [comment]
:+1:

**Comment by nzakas** [comment]
@keithamus I'd prefer `always-multiline` to make it clear what applies to multiline. Otherwise, +1.

**Comment by keithamus** [comment]
@nzakas @UltCombo See #2000 :smile:

**Comment by steadicat** [comment]
Shouldn't this be valid with "always-multiline"?
```
{x: {
  foo: 'bar',
}}
```

**Comment by UltCombo** [comment]
@steadicat yes, this was my initial idea as well (see https://github.com/eslint/eslint/issues/1984#issuecomment-77985750).
I don't particularly tend to write multiple properties in the same line when the literal takes more than a single line, but that approach would probably be better overall.

**Comment by keithamus** [comment]
@steadicat can I recommend raising a new issue?

**Comment by nzakas** [comment]
Yes, please do. Commenting on closed issues makes conversations hard to track.
